opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -1241,7 +1245,141 @@ namespace Opm {
+
+                auto mismatch = [&] (auto well_group_thp) {
+                    Scalar well_group_rate(0.0);
+                    Scalar rate(0.0);
+                    for (auto& well : this->well_container_) {
+                        std::string well_name = well->name();
+                        auto& ws = well_state.well(well_name);
+                        if (group.hasWell(well_name)) {
+                            well->setDynamicThpLimit(well_group_thp);
+                            const Well& well_ecl = this->wells_ecl_[well->indexOfWell()];
+                            const auto inj_controls = Well::InjectionControls(0);
+                            const auto prod_controls = well_ecl.productionControls(summary_state);
+                            well->iterateWellEqWithSwitching(this->simulator_, dt, inj_controls, prod_controls, well_state, group_state, local_deferredLogger,  false, false);
+                            rate = -tcalc.calcModeRateFromRates(ws.surface_rates);
+                            well_group_rate += rate;
+                        }
+                    }
+                    return (well_group_rate - orig_target)/orig_target;
+                };
 
+                Scalar min_thp, max_thp;
+                std::array<Scalar, 2> range_initial;
+                //Find an initial bracket
+                if (!this->well_group_thp_calc_.has_value()){
+                    // Retrieve the terminal pressure of the associated root of the manifold group
+                    std::string node_name =  nodeName;
+                    while (!network.node(node_name).terminal_pressure().has_value()) {
+                        auto branch = network.uptree_branch(node_name).value();
+                        node_name = branch.uptree_node();
+                    }
+
+                    min_thp = network.node(node_name).terminal_pressure().value();
+                    WellBhpThpCalculator<Scalar>::bruteForceBracketCommonTHP(mismatch, min_thp, max_thp);
+                    // Narrow down the bracket
+                    Scalar low1, high1;
+                    std::array<Scalar, 2> range = {0.9*min_thp, 1.1*max_thp};
+                    std::optional<Scalar> appr_sol;
+                    WellBhpThpCalculator<Scalar>::bruteForceBracketCommonTHP(mismatch, range, low1, high1, appr_sol, 0.0, local_deferredLogger);
+                    min_thp = low1;
+                    max_thp = high1;
+                    range_initial = {min_thp, max_thp};
+                }
+
+                const auto upbranch = network.uptree_branch(nodeName);
+                const auto it = this->node_pressures_.find((*upbranch).uptree_node());
+                const Scalar nodal_pressure = it->second;
+                Scalar well_group_thp = nodal_pressure;
+
+                if (!this->well_group_thp_calc_.has_value() || this->well_group_thp_calc_ > nodal_pressure) {
```

Comment:
This function is trying to calculate `this->well_group_thp_calc_`. With the condition `this->well_group_thp_calc_ > nodal_pressure`, will it happens that, when ``this->well_group_thp_calc_ > nodal_pressure` happens, we will never update `this->well_group_thp_calc_` anymore? At least, from the code, it can happens. Not sure it is a desired behavoir.

Reply:
I have tried to remove this (`this->well_group_thp_calc_ > nodal_pressure`) condition which goes with additional calculation costs but would be more general. However oscillating group thp pressure results (figure 1).  Tightening the tolerance  (`--tolerance-wells=1E-6`) gives smoother results but shut-in of wells (figure 2).  Compare both figures with the one given earlier (see above). Can we have a chat about what to do here?

figure 1:
![image](https://github.com/OPM/opm-simulators/assets/56635911/e0da18f6-8b2a-4c42-86e0-58658de77677)

figure 2:
![image](https://github.com/OPM/opm-simulators/assets/56635911/c09e7a32-b7ed-4a6f-8b30-7d9df72759bf)

Reply:
Still for information and potential discussion, I think if with tighter tolerance, only well B-1H got shut, which is the behavoir from the reference simulator, I think it is fine.

Reply:
The oscillation looks like solution is not converged.

Reply:
But let me check why the well gets SHUT.

Reply:
The manifold THP can not be lower than the nodal pressure. Ideally, we can check whether we can determine whether we can reach the production target first. If not, then the wells will run under the nodal pressure, so we do not need to look for the manifold pressure. 

If we can reach the production target, then we need to calculate the manifold choke pressure. 

But determining whether we can reach the production target involving solving all the wells under THP constraints might fail for various causes. 

The current approach assumes that we use the existing autochoke_thp (solved previously or equal to the nodal pressure ) to determine whether the group production can be reached. It works when the situation will not sway back and force between being able and not being able to each the group target.  It is the situation for the case the PR has been tested with.  The situations for other cases can be very different.